### PR TITLE
NIAD-616: order by as codeable concept

### PR DESCRIPTION
**Pull request by bartek-sarul**

### CODE REVIEW

**Review by Broom94 — APPROVED** [review]

**Comment by dawidskainos** [comment]
I think it is not enough to add OrderedBy:eventDate. We should also sort list entries by date - where possible.

File: service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/comparator/ResourceComparator.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+        if (date1 == null ^ date2 == null) {
+            return (date1 == null) ? -1 : 1;
+        }
+
+        return date1.compareTo(date2);
+    }
+
+    private Date getResourceDate(Resource resource) {
+        Date date = null;
```
Comment (dawidskainos) [inline service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/comparator/ResourceComparator.java:43]: Is the purpose of this variable to be always null? ;)

File: service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/mapper/ListMapper.java [hunk 2]
```diff
@@ -58,17 +58,22 @@ public ListResource mapList(POCDMT000002UK01ClinicalDocument1 clinicalDocument,
...
             .setDate(new Date())
             .setSource(TRANSFORMER_DEVICE)
-            .setOrderedBy(new CodeableConcept().setText(ORDER_BY));
+            .setOrderedBy(createOrderByConcept());
+
+        List<Resource> resourcesCreatedOrderedList = new ArrayList();
 
-        resourcesCreated.stream()
-            .filter(it -> TRIAGE_RESOURCES.contains(it.getClass()))
+        resourcesCreated.stream().sequential().collect(Collectors.toCollection(() -> resourcesCreatedOrderedList));
```
Comment (dawidskainos) [inline service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/mapper/ListMapper.java:72]: why are making a copy of the resourcesCreated list?

File: service/src/test/java/uk/nhs/adaptors/oneoneone/cda/report/comparator/ResourceComparatorTest.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+import org.hl7.fhir.dstu3.model.ReferralRequest;
+import org.hl7.fhir.dstu3.model.ResourceType;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+@ExtendWith(MockitoExtension.class)
+public class ResourceComparatorTest {
```
Comment (dawidskainos) [inline service/src/test/java/uk/nhs/adaptors/oneoneone/cda/report/comparator/ResourceComparatorTest.java:17]: It's ok to use Mockito here, however it looks to me like an overkill. You can very easily create resources and set necessary dates without using mocking tools.

File: service/src/test/java/uk/nhs/adaptors/oneoneone/cda/report/comparator/ResourceComparatorTest.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    @InjectMocks
+    private ResourceComparator resourceComparator;
+    @Mock
+    private ReferralRequest referralRequest1;
+    @Mock
+    private ReferralRequest referralRequest2;
+
+    @Test
+    public void shouldReturnNegativeOneForValidDates() {
```
Comment (dawidskainos) [inline service/src/test/java/uk/nhs/adaptors/oneoneone/cda/report/comparator/ResourceComparatorTest.java:26]: All these tests check comparator functionality pretty well, however they're hard to understand without digging into the code.

What I'd do:

    @Test
    public void shouldReturnOneWhenFirstResourceIsLater() {
        ReferralRequest first = new ReferralRequest().setAuthoredOn(LATER_DATE);
        ReferralRequest second = new ReferralRequest().setAuthoredOn(EARLIER_DATE);

        assertThat(resourceComparator.compare(first, second)).isEqualTo(1);
    }

**Review by dawidskainos — CHANGES_REQUESTED** [review]

**Review by saddiqs1 — APPROVED** [review]

**Review by Broom94 — APPROVED** [review]

**Pull request merged**